Blank import name in unresolved import warning, hematite, Rust 1.15

https://github.com/PistonDevelopers/hematite 6a7bb8d067f9144003be0f629cab0004f2380535

```
brian@ip-10-145-43-250:/mnt2/dev⟫ rustc +beta -Vv
rustc 1.15.0-beta.3 (a035041ba 2017-01-07)
binary: rustc
commit-hash: a035041ba450ce3061d78a2bdb9c446eb5321d0d
commit-date: 2017-01-07
host: x86_64-unknown-linux-gnu
release: 1.15.0-beta.3
LLVM version: 3.9
```

```
   Compiling piston-gfx_texture v0.18.0
   Compiling piston3d-gfx_voxel v0.12.0
   Compiling hematite v0.0.12 (file:///mnt2/dev/hematite)
warning: `$crate` may not be imported
  --> src/shader.rs:39:1
   |
39 |   gfx_pipeline!( pipe {
   |  _^ starting here...
40 | |     vbuf: ::gfx::VertexBuffer<Vertex> = (),
41 | |     transform: ::gfx::Global<[[f32; 4]; 4]> = "u_projection",
42 | |     view: ::gfx::Global<[[f32; 4]; 4]> = "u_view",
43 | |     color: ::gfx::TextureSampler<[f32; 4]> = "s_texture",
44 | |     out_color: ::gfx::RenderTarget<::gfx::format::Srgba8> = "out_color",
45 | |     out_depth: ::gfx::DepthTarget<::gfx::format::DepthStencil> =
46 | |         ::gfx::preset::depth::LESS_EQUAL_WRITE,
47 | | });
   | |___^ ...ending here
   |
   = note: `use $crate;` was erroneously allowed and will become a hard error in a future release
   = note: this error originates in a macro outside of the current crate

error[E0432]: unresolved import ``
  --> src/shader.rs:39:1
   |
39 |   gfx_pipeline!( pipe {
   |  _^ starting here...
40 | |     vbuf: ::gfx::VertexBuffer<Vertex> = (),
41 | |     transform: ::gfx::Global<[[f32; 4]; 4]> = "u_projection",
42 | |     view: ::gfx::Global<[[f32; 4]; 4]> = "u_view",
43 | |     color: ::gfx::TextureSampler<[f32; 4]> = "s_texture",
44 | |     out_color: ::gfx::RenderTarget<::gfx::format::Srgba8> = "out_color",
45 | |     out_depth: ::gfx::DepthTarget<::gfx::format::DepthStencil> =
46 | |         ::gfx::preset::depth::LESS_EQUAL_WRITE,
47 | | });
   | |___^ ...ending here: no `` in the root
   |
   = note: this error originates in a macro outside of the current crate

error: aborting due to previous error

error: Could not compile `hematite`.

To learn more, run the command again with --verbose.
```

Notice that the name of the import is blank.

__Interestingly__ on nightly this is not an error, but a warning, and the import is not blank but '$':

```
warning: unused import: `$`, #[warn(unused_imports)] on by default
  --> src/shader.rs:39:1
   |
```

I guess it would be best if beta had the same behavior as nightly.

cc @jseyfried

--- a/src/librustc_resolve/build_reduced_graph.rs
+++ b/src/librustc_resolve/build_reduced_graph.rs
@@ -143,7 +143,7 @@ impl<'a> Resolver<'a> {
                 let is_prelude = attr::contains_name(&item.attrs, "prelude_import");
 
                 match view_path.node {
-                    ViewPathSimple(binding, ref full_path) => {
+                    ViewPathSimple(mut binding, ref full_path) => {
                         let mut source = full_path.segments.last().unwrap().identifier;
                         let source_name = source.name;
                         if source_name == "mod" || source_name == "self" {
@@ -157,6 +157,9 @@ impl<'a> Resolver<'a> {
                                 ModuleKind::Block(..) => unreachable!(),
                             };
                             source.name = crate_name;
+                            if binding.name == "$crate" {
+                                binding.name = crate_name;
+                            }
 
                             self.session.struct_span_warn(item.span, "`$crate` may not be imported")
                                 .note("`use $crate;` was erroneously allowed and \

--- a/src/test/compile-fail/auxiliary/import_crate_var.rs
+++ b/src/test/compile-fail/auxiliary/import_crate_var.rs
@@ -8,5 +8,10 @@
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
 
+pub fn f() {}
+
 #[macro_export]
-macro_rules! m { () => { use $crate; } }
+macro_rules! m { () => {
+    use $crate;
+    import_crate_var::f();
+} }

--- a/src/test/compile-fail/import-crate-var.rs
+++ b/src/test/compile-fail/import-crate-var.rs
@@ -11,11 +11,13 @@
 // aux-build:import_crate_var.rs
 // error-pattern: `$crate` may not be imported
 // error-pattern: `use $crate;` was erroneously allowed and will become a hard error
+// error-pattern: compilation successful
 
 #![feature(rustc_attrs)]
 
 #[macro_use] extern crate import_crate_var;
-m!();
 
 #[rustc_error]
-fn main() {}
+fn main() {
+    m!();
+}

